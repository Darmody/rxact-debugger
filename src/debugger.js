// @flow
import { StateStream } from 'rxact'
import {
  operatorLogger,
  emitterLogger,
} from './logger'
import {
  isOperator,
  isEmitter,
  isDispose,
  isDebugger,
  isPresent,
} from './helpers'

export const clean = () => {
  window.rxactDebugger = undefined
}

function createDebuggerProxy(instance) {
  const streamName = instance.streamName
  const handler = {
    get: (target, prop) => {
      if (isDebugger(prop)) {
        return {
          start: () => {
            this.isStarted = true
          },
          stop: () => {
            this.isStarted = false
          }
        }
      }

      let type = 'operator'

      if (!isPresent(target, prop)) {
        console.warn(`Operator does not exist: ${streamName}.${prop}`)
      }
      if (!isOperator(target, prop) || isEmitter(target.emitters, prop)) {
        return target[prop]
      }

      if (isDispose(prop)) {
        type = 'dispose'
      }

      return (...params) => {
        if (this.isStarted) {
          operatorLogger(type, 'log', streamName, prop)
        }

        return target[prop](...params)
      }
    },
    set: (target, prop, value) => {
      if (target[prop]) {
        console.warn(`Operator redefined: ${streamName}.${prop}`)
      }
      target[prop] = value

      return true
    },
  }

  return new Proxy(instance, handler)
}

function createEmitterProxy(instance) {
  const handler = {
    apply: (target, thisArg, argumentsList) => {
      const [name, emitter] = argumentsList

      const loggableEmitter = (...params) => prevState => {
        const nextState = emitter(...params)(prevState)

        if (this.isStarted) {
          emitterLogger(instance.streamName, name, prevState, nextState)
        }

        return nextState
      }

      return target.call(thisArg, name, loggableEmitter)
    }
  }

  return new Proxy(instance.emitter, handler)
}

export type Options = {
  start?: Boolean | Array<string>,
}

const isStartDefault = (start, stream) => start === true || (Array.isArray(start) &&
  start.find(item => item === stream))

const defaultOptions = {
  start: false,
}

const rxactDebugger = (options: Options = {}) =>  {
  if (typeof options !== 'object' || Array.isArray(options)) {
    throw new Error('rxact-debugger: Expect options to be an object.')
  }

  return (instance: StateStream) => {
    const finalOptions = { ...defaultOptions, ...options }
    const { start } = finalOptions
    const startDefault = isStartDefault(start, instance.streamName)
    const context = {
      isStarted: false
    }

    const debuggableProxy = createDebuggerProxy.call(context, instance)
    const emitterProxy = createEmitterProxy.call(context, instance)

    instance.emitter = emitterProxy

    if (!window.rxactDebugger) {
      window.rxactDebugger = {}
    }
    window.rxactDebugger[instance.streamName] = debuggableProxy

    if (startDefault) {
      debuggableProxy.debugger.start()
    }

    return debuggableProxy
  }
}

export default rxactDebugger
