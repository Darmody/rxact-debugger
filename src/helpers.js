// @flow
import type { StateStream } from 'rxact'

type Target = StateStream
type Emitters = {}
type Prop = string

export const METHOD_BLACK_LIST = [
  'constructor',
  'next',
  'eventRunner',
  'getState',
  'emitter',
]

export const isOperator = (target: Target, prop: Prop) =>
  typeof target[prop] === 'function' && !METHOD_BLACK_LIST.find(method => method === prop)
export const isEmitter = (emitters: Emitters, prop: Prop) => !!emitters[prop]
export const isDispose = (prop: Prop) => prop === 'dispose'
export const isDebugger = (prop: Prop) => prop === 'debugger'
export const isPresent = (target: Target, prop: Prop) => !!target[prop]
