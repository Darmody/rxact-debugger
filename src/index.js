// @flow
import type { StateStream } from 'rxact'
import type { Options } from './debugger'
import rxactDebugger, { clean } from './debugger'

const plugin = rxactDebugger
const decorator = (options: Options) => {
  const debuggerPlugin = rxactDebugger(options)

  return (StateStream: StateStream) => {
    return class DebuggableStream extends StateStream {
      constructor(...params: Array<any>) {
        const instance = super(...params)

        return debuggerPlugin(instance)
      }
    }
  }
}

export {
  plugin,
  decorator,
  clean,
}
