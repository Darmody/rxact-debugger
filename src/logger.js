// @flow
const BOLD = 'font-weight:bold;'
const UNDERLINE = 'text-decoration:underline;'
const COLOR = color => `color:${color};`

const toObject = (state) => {
  if (state.toJS) {
    return state.toJS()
  }

  return state
}

export const operatorLogger = (
  type: string, logger: string = 'log', streamName: string, operatorName: string
) => {
  let prefix = ''
  let action = ''
  let color = ''

  switch (type) {
    case 'emitter': {
      prefix = 'STATE   '
      action = 'updated'
      color = '#3dbd7d'
      break
    }
    case 'dispose': {
      prefix = 'DISPOSE   '
      action = 'disposed'
      color = '#f04134'
      break
    }
    default: {
      prefix = 'OPERATOR'
      action = 'called '
      color = '#948aec'
      break
    }
  }

  console[logger](
    `%c${prefix} %c ${action}  %c${streamName}->${operatorName}`,
    BOLD + COLOR(color),
    BOLD + COLOR('#108ee9'),
    UNDERLINE + BOLD + COLOR('#5a5a5a'),
  )
}

const stateLog = (color, content, state) => {
  console.log(
    `%c${content}`,
    COLOR(color) + BOLD,
    state,
  )
}

export const emitterLogger = (
  streamName: string,
  emitterName: string,
  prevState: any,
  nextState: any,
) => {
  operatorLogger('emitter', 'groupCollapsed', streamName, emitterName)
  stateLog('#5a5a5a', 'prevState:', toObject(prevState))
  stateLog('#ffce3d', 'nextState:', toObject(nextState))
  console.groupEnd()
}
