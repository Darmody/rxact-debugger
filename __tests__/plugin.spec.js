import { StateStream } from 'rxact'
import { plugin } from '../src/index'
import debuggerExample from './shared/debugger.example'
import optionExample from './shared/option.example'

describe('plugin', () => {
  debuggerExample(StateStream, [plugin()])
  optionExample(plugin, () => Component => Component)
})
