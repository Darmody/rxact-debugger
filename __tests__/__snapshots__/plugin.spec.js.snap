// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin Debugger dispose logging 1`] = `
Array [
  Array [
    "%cDISPOSE    %c disposed  %cstream->dispose",
    "font-weight:bold;color:#f04134;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
]
`;

exports[`plugin Debugger emitter logging 1`] = `
Array [
  Array [
    "%cOPERATOR %c called   %cstream->emitter1",
    "font-weight:bold;color:#948aec;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
  Array [
    "%cprevState:",
    "color:#5a5a5a;font-weight:bold;",
    0,
  ],
  Array [
    "%cnextState:",
    "color:#ffce3d;font-weight:bold;",
    1,
  ],
]
`;

exports[`plugin Debugger emitter logging 2`] = `
Array [
  Array [
    "%cSTATE    %c updated  %cstream->emitter1",
    "font-weight:bold;color:#3dbd7d;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
]
`;

exports[`plugin Debugger operator does not exist 1`] = `
Array [
  Array [
    "Operator does not exist: stream.operator",
  ],
]
`;

exports[`plugin Debugger operator logging 1`] = `
Array [
  Array [
    "%cOPERATOR %c called   %cstream->operator1",
    "font-weight:bold;color:#948aec;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
]
`;

exports[`plugin Debugger operator redefine 1`] = `
Array [
  Array [
    "Operator redefined: stream.operator",
  ],
]
`;

exports[`plugin Debugger stop logging 1`] = `
Array [
  Array [
    "%cprevState:",
    "color:#5a5a5a;font-weight:bold;",
    0,
  ],
  Array [
    "%cnextState:",
    "color:#ffce3d;font-weight:bold;",
    1,
  ],
]
`;

exports[`plugin Debugger stop logging 2`] = `
Array [
  Array [
    "%cSTATE    %c updated  %cstream->emitter1",
    "font-weight:bold;color:#3dbd7d;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
]
`;

exports[`plugin Debugger work with Immutable 1`] = `
Array [
  Array [
    "%cprevState:",
    "color:#5a5a5a;font-weight:bold;",
    Object {
      "state": 0,
    },
  ],
  Array [
    "%cnextState:",
    "color:#ffce3d;font-weight:bold;",
    Object {
      "state": 0,
    },
  ],
]
`;

exports[`plugin options start loggin on all streams 1`] = `
Array [
  Array [
    "%cprevState:",
    "color:#5a5a5a;font-weight:bold;",
    0,
  ],
  Array [
    "%cnextState:",
    "color:#ffce3d;font-weight:bold;",
    1,
  ],
]
`;

exports[`plugin options start loggin on all streams 2`] = `
Array [
  Array [
    "%cSTATE    %c updated  %cstream->emitter1",
    "font-weight:bold;color:#3dbd7d;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
]
`;

exports[`plugin options start logging on part streams 1`] = `
Array [
  Array [
    "%cprevState:",
    "color:#5a5a5a;font-weight:bold;",
    0,
  ],
  Array [
    "%cnextState:",
    "color:#ffce3d;font-weight:bold;",
    1,
  ],
]
`;

exports[`plugin options start logging on part streams 2`] = `
Array [
  Array [
    "%cSTATE    %c updated  %ctest1->emitter1",
    "font-weight:bold;color:#3dbd7d;",
    "font-weight:bold;color:#108ee9;",
    "text-decoration:underline;font-weight:bold;color:#5a5a5a;",
  ],
]
`;
